[1.19] Menus and Screens: An HCI Research Paper (#465)
MinecraftForge/Documentation: docs/gui/screens.md

Second, there is the `#fill` method, which draws a rectangle to the screen. The line methods internally call this method. This takes in the left x coordinate, the top y coordinate, the right x coordinate, the bottom y coordinate, and the color.

Finally, there is the `#fillGradient` method, which draws a rectangle with a vertical gradient. This takes in the right x coordinate, the bottom y coordinate, the left x coordinate, the top y coordinate, the z coordinate, and the bottom and top colors.

### Strings

Strings are drawn through its `Font`, typically consisting of their own shaders for normal, see through, and offset mode. There are two alignment of strings that can be rendered, each with a back shadow: a left-aligned string (`#drawString`) and a center-aligned string (`#drawCenteredString`). These both take in the font the string will be rendered in, the string to draw, the x coordinate representing the left or center of the string respectively, the top y coordinate, and the color.

!!! note
    Strings should typically be passed in as [`Component`s][component] as they handle a variety of usecases, including the two other overloads of the method.

### Textures

> Textures are drawn through blitting, hence the method name `#blit`, which, for this purpose, copies the bits of an image and draws them directly to the screen. These are drawn through a position texture shader. While there are many different `#blit` overloads, we will only discuss two: the instance `#blit` and the static `#blit` the instance method delegates to.

Reviewer: ```suggestion
Textures are drawn through blitting, hence the method name `#blit`, which, for this purpose, copies the bits of an image and draws them directly to the screen. These are drawn through a position texture shader. While there are many different `#blit` overloads, we will only discuss two: the instance `#blit` and the static `#blit` which the instance method delegates to.
```

Some common widgets which may be used are `Button`, its subtype `ImageButton`, and `EditBox` which is used to input text on the screen.

## GuiEventListener

Any screen rendered in Minecraft implements `GuiEventListener`. `GuiEventListener`s are responsible for handling user interaction with the screen. These include inputs from the mouse (movement, clicked, released, dragged, scrolled, mouseover) and keyboard (pressed, released, typed). Each method returns whether the associated action affected the screen successfully. Widgets like buttons, chat boxes, lists, etc. also implement this interface.

### ContainerEventHandler

Almost synonymous with `GuiEventListener`s are their subtype: `ContainerEventHandler`s. These are responsible for handling user interaction on screens which contain widgets, managing which is currently focused and how the associated interactions are applied. `ContainerEventHandler`s add three additional features: interactable children, dragging, and focusing.

> Event handlers held children which was used to determine the interaction order of elements. During the mouse event handlers (excluding dragging), the first child in the list that the mouse was over would have their logic executed.

Reviewer: This paragraph seems a bit... off. At the very least, the first sentence has an odd choice of verb tense. We should use either the present or future tense for verbs, like so:

```suggestion
Event handlers hold children which are used to determine the interaction order of elements. During the mouse event handlers (excluding dragging), the first child in the list that the mouse was hovering over will have their logic executed.
```

Dragging is implemented within `#mouseClicked` and `#mouseReleased` allowing for logic to be more precisely executed when the mouse was dragging the element.

> Focusing allowed for a specific child to be selected to execute logic. These included during keyboard events and when the mouse was being dragged. Focus of which element was typically set through `#setFocused` or, when the screen was being opened, `#setInitialFocus`. In addition, interactable children could be cycled through using `#changeFocus`, selecting the next child in the list, or the previous child when shift was held down.

Reviewer: These also seem to have some odd choices for verb tenses.

    // Assume we have some Component 'label'
    // 'label' is drawn at 'labelX' and 'labelY'
    this.font.draw(pose, label, labelX, labelY, 0x404040);
}
```

!!! note
    When rendering the label, you do **not** need to specify the `leftPos` and `topPos` offset. Those have already been translated within the `PoseStack` so everything within this method is drawn relative to those coordinates.

## Registering an AbstractContainerScreen

To use an `AbstractContainerScreen` with a menu, it needs to be registered. This can be done by calling `MenuScreens#register` within the `FMLClientSetupEvent` on the [**mod event bus**][modbus].

```java
// Event is listened to on the mod event bus
private void clientSetup(FMLClientSetupEvent event) {
    event.enqueueWork(
        // Assume RegistryObject<MenuType<MyMenu>> MY_MENU
        // Assume MyContainerScreen<MyMenu> which takes in three parameters
        () -> MenuScreens.register(MY_MENU.get(), MyContainerScreen::new)
    );
}
```

!!! warning
    `MenuScreens#register` is not thread-safe, so it needs to be called inside `#enqueueWork` provided by the parallel dispatch event.

[menus]: ./menus.md
[network]: ../networking/index.md

> [screen]: #TODO

Reviewer: `#TODO` :shipit:

Almost synonymous with `GuiEventListener`s are their subtype: `ContainerEventHandler`s. These are responsible for handling user interaction on screens which contain widgets, managing which is currently focused and how the associated interactions are applied. `ContainerEventHandler`s add three additional features: interactable children, dragging, and focusing.

Event handlers hold children which are used to determine the interaction order of elements. During the mouse event handlers (excluding dragging), the first child in the list that the mouse hovers over has their logic executed.

Dragging an element with the mouse, implemented via `#mouseClicked` and `#mouseReleased`, provides more precisely executed logic.

> Focusing allows for a specific child to be checked first and handled during an event's execution, such during keyboard events or dragging the mouse. Focus is typically set through `#setFocused` or, when the screen is being opened, `#setInitialFocus`. In addition, interactable children can be cycled using `#changeFocus`, selecting the next child in the list, or the previous child if the shift key is held down.

Reviewer: Missing `as` in `such during`:
```suggestion
Focusing allows for a specific child to be checked first and handled during an event's execution, such as during keyboard events or dragging the mouse. Focus is typically set through `#setFocused` or, when the screen is being opened, `#setInitialFocus`. In addition, interactable children can be cycled using `#changeFocus`, selecting the next child in the list, or the previous child if the shift key is held down.
```
